### PR TITLE
#325 add State Store configs to support stateful functions in Pulsar 

### LINKED ISSUE #325: support stateful function
- add stateful function related configs
- integrate into FunctionDetails
- add tests
- add docs

### PR DESCRIPTION
**Pull request by freeznet**
Fixes #325

### Modifications

- [x] Add `Stateful` Resource to support config stateful function's state store backend. 
- [x] Add `PulsarStateStore` Resource to provide state config for pulsar functions.
- [x] Add `--state_storage_serviceurl` and `--state_storage_impl_class` support for pulsar functions.
- [x] Add tests
- [x] Add examples

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Documentation

Check the box below.

Need to update docs? 

- [x] `doc-required` 
  
  https://github.com/streamnative/function-mesh/issues/360
  
- [ ] `no-need-doc` 
  
  (Please explain why)
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

File: api/v1alpha1/function_types.go [hunk 1]
```diff
@@ -85,6 +85,9 @@ type FunctionSpec struct {
 
 	// Image pull policy, one of Always, Never, IfNotPresent, default to IfNotPresent.
 	ImagePullPolicy corev1.PullPolicy `json:"imagePullPolicy,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	StatefulFunction *Stateful `json:"statefulFunction,omitempty"`
```
Comment (nlu90) [inline api/v1alpha1/function_types.go:90]: how about `StateConfig` ?

File: api/v1alpha1/common.go [hunk 2]
```diff
@@ -54,6 +58,25 @@ type PulsarMessaging struct {
...
+}
+
+type PulsarStateStoreJavaProvider struct {
+	// The java class name of the state store provider implementation
+	// The class must implement `org.apache.pulsar.functions.instance.state.StateStoreProvider` interface
+	// If not set, `org.apache.pulsar.functions.instance.state.BKStateStoreProviderImpl` will be used
+	ClassName string `json:"className"`
+
+	// The configmap of the configuration for the state store provider
+	Config string `json:"config,omitempty"`
```
Comment (nlu90) [inline api/v1alpha1/common.go:77]: should use `map[string]string`?

File: api/v1alpha1/validate.go [hunk 3]
```diff
@@ -290,3 +311,23 @@ func validateAutoAck(autoAck *bool) *field.Error {
 	}
 	return nil
 }
+
+func validateStatefulFunctionConfigs(statefulFunctionConfigs *Stateful, runtime Runtime) *field.Error {
+	if statefulFunctionConfigs != nil {
+		if statefulFunctionConfigs.Pulsar != nil {
+			if isGolangRuntime(runtime) {
```
Comment (nlu90) [inline api/v1alpha1/validate.go:318]: does python runtime support stateful feature?

File: api/v1alpha1/common.go (hunk 2 above)
Comment (freeznet) [reply api/v1alpha1/common.go:77]: use `*Config` instead, since `*Config` supports `map[string]interface{}`

File: api/v1alpha1/validate.go (hunk 3 above)
Comment (freeznet) [reply api/v1alpha1/validate.go:318]: yes, python support stateful, see https://github.com/apache/pulsar/blob/master/pulsar-functions/instance/src/main/python/python_instance.py#L338-L343

**Pull request merged**